src/psyclone/psyir/transformations/intrinsics/dotproduct2code_trans.py

```diff
@@ -0,0 +1,321 @@
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
+
+'''Module providing a transformation from a PSyIR DOT_PRODUCT operator
+to PSyIR code. This could be useful if the DOT_PRODUCT operator is not
+supported by the back-end or if the performance in the inline code is
+better than the intrinsic.
+
+'''
+from __future__ import absolute_import
+from psyclone.psyir.nodes import BinaryOperation, Assignment, Reference, \
+    Loop, Literal, ArrayReference, Range, Routine
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE, REAL_TYPE, \
+    ArrayType, ScalarType
+from psyclone.psyir.transformations import TransformationError
+from psyclone.psyir.transformations.intrinsics.operator2code_trans import \
+    Operator2CodeTrans
+
+
+def _get_array_bound(vector1, vector2, writer):
+    '''A utility function that returns the appropriate loop bounds (lower,
+    upper and step) for a vector.
+    If either of the vectors are declared with known bounds (an integer or a
+    symbol) then these bound values are used. If the size is unknown
+    (a deferred or attribute type) then the LBOUND and UBOUND PSyIR
+    nodes are used.
+
+    :param array: the reference that we are interested in.
+    :type array: :py:class:`psyir.nodes.Reference`
+    :param int index: the (array) reference index that we are \
+        interested in.
+
+    :returns: the loop bounds for this array index.
+    :rtype: (Literal, Literal, Literal) or \
+        (BinaryOperation, BinaryOperation, Literal)
+
+    :raises TransformationError: if both vector1 and vector2 are not \
+        references to array symbols.
+
+    '''
+    if not ((isinstance(vector1, Reference) and vector1.symbol.is_array) or
+            (isinstance(vector2, Reference) and vector2.symbol.is_array)):
+        raise TransformationError(
+            f"DotProduct2CodeTrans._get_array_bound requires at least one of "
+            f"the dotproduct arguments to be an array but found "
+            f"'{writer(vector1)}' and '{writer(vector2)}'.")
+
+    # Look for explicit bounds in one of the array declarations
+    for vector in [vector1, vector2]:
+        symbol = vector.symbol
+        my_dim = symbol.shape[0]
+        if isinstance(my_dim, ArrayType.ArrayBounds):
+            lower_bound = my_dim.lower
+            upper_bound = my_dim.upper
+            step = Literal("1", INTEGER_TYPE)
+            return (lower_bound, upper_bound, step)
+
+    # No explicit array bound information could be found for either
+    # array so use the LBOUND and UBOUND intrinsics.
+    symbol = vector1.symbol
+    my_dim = symbol.shape[0]
+    lower_bound = BinaryOperation.create(
+        BinaryOperation.Operator.LBOUND, Reference(symbol),
+        Literal("1", INTEGER_TYPE))
```

Comment:
See earlier note - we can't assume that the dimension is number 1.

Reply:
As mentioned earlier, the validate method ensures this and I've added some information to the `_get_array_bound()` function to explain this.